plugins/discourse-local-dates/assets/javascripts/lib/rich-editor-extension.js

```diff
@@ -0,0 +1,188 @@
+          getAttrs: (dom) => {
+            return {
+              fromDate: dom.dataset.fromDate,
+              toDate: dom.dataset.toDate,
+              fromTime: dom.dataset.fromTime,
+              toTime: dom.dataset.toTime,
+              timezone: dom.dataset.timezone,
+            };
+          },
+        },
+      ],
+      toDOM: (node) => {
+        const fromTimeStr = node.attrs.fromTime
+          ? ` ${node.attrs.fromTime}`
+          : "";
+        const toTimeStr = node.attrs.toTime ? ` ${node.attrs.toTime}` : "";
+        return [
+          "span",
+          { class: "discourse-local-date-wrapper" },
+          [
+            "span",
+            {
+              class: "discourse-local-date cooked-date",
+              "data-range": "from",
+              "data-date": node.attrs.fromDate,
+              "data-time": node.attrs.fromTime,
+              "data-timezone": node.attrs.timezone,
+            },
+            `${node.attrs.fromDate}${fromTimeStr}`,
+          ],
+          " → ",
+          [
+            "span",
+            {
+              class: "discourse-local-date cooked-date",
+              "data-range": "to",
+              "data-date": node.attrs.toDate,
+              "data-time": node.attrs.toTime,
+              "data-timezone": node.attrs.timezone,
+            },
+            `${node.attrs.toDate}${toTimeStr}`,
+          ],
+        ];
+      },
+    },
+  },
+  parse: {
+    span_open(state, token) {
+      if (token.attrGet("class") !== "discourse-local-date") {
+        return;
+      }
+
+      if (token.attrGet("data-range") === "from") {
+        state.openNode(state.schema.nodes.local_date_range, {
+          fromDate: token.attrGet("data-date"),
+          fromTime: token.attrGet("data-time"),
+          timezone: token.attrGet("data-timezone"),
+        });
+        state.__localDateRange = true;
+        return true;
+      }
+
+      if (token.attrGet("data-range") === "to") {
+        // In our markdown-it tokens, a range is a series of span_open/span_close/span_open/span_close
+        // We skip opening a node for `to` and set it on the top node
+        state.top().attrs.toDate = token.attrGet("data-date");
+        state.top().attrs.toTime = token.attrGet("data-time");
+        delete state.__localDateRange;
+        return true;
+      }
+
+      state.openNode(state.schema.nodes.local_date, {
+        date: token.attrGet("data-date"),
+        time: token.attrGet("data-time"),
+        timezone: token.attrGet("data-timezone"),
+      });
+      return true;
+    },
+    span_close(state) {
+      if (["local_date", "local_date_range"].includes(state.top().type.name)) {
+        if (!state.__localDateRange) {
+          state.closeNode();
+        }
+        return true;
+      }
+    },
+  },
+  serializeNode({ utils: { isBoundary } }) {
+    return {
+      local_date(state, node, parent, index) {
+        if (!isBoundary(state.out, state.out.length - 1)) {
```

Comment:
We've been using this to force a space between some strings and their surrounds. In this case we're also supporting `PluginParams` in `serializeNode` and `serializeMark`, so we can re-use this without needing to import directly from within `/static/`.

That said, another option (if we care about this in the future...) could be enforcing this **in the ProseMirror document** itself.